onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h

```diff
@@ -314,33 +312,50 @@ class DnnlConvGrad : public DnnlKernel {
             *conv_bwd_weights_desc_, engine_to_use, *(conv_fwd_->GetPrimitiveDesc())));
 
     diff_dst_mem_ = onnxruntime::make_unique<dnnl::memory>(
-        dnnl::memory(conv_bwd_weights_pd_.get()->diff_dst_desc(), engine_to_use, nullptr));
+        dnnl::memory(conv_bwd_weights_pd_.get()->diff_dst_desc(), cpu_engine, nullptr));
```

Comment:
Checking my understanding here: is it intentional that these will be created for the CPU engine even when a GPU is available?

(I would have expected e.g. `diff_dst_mem_` to contain a reference to GPU memory when a GPU is available, and the same member to point to CPU memory otherwise. This seems to be how `primitive_dst_mem_` is handled. But I'm not familiar with how this is actually implemented....)

Reply:
This is describing the memory block that will be bound to in the Bind function. This is memory is from OnnxRuntime and exists in the CPU ram, not the GPU ram.  This memory will be copied to the GPU using the reorder below (Lines 327-328). So the cpu_engine is done purposely and is required.

Reply:
Thanks for clarifying!